File: src/utils/secretStore.ts
/**
 * 密钥安全存储 - 优先走系统 Keychain，浏览器环境降级到 XOR 混淆
 *
 * Tauri 桌面环境：
 *   - Windows: DPAPI（Credential Manager）
 *   - macOS:   Keychain
 *   - Linux:   Secret Service
 *
 * 浏览器开发环境（npm run dev 而非 tauri dev）：
 *   - 降级到 localStorage + XOR 混淆（仅作开发体验用，仍非密码学安全）
 *
 * 桌面写入失败明确报错，不静默降级；旧本地副本可读取，成功保存后清理。
 */

import { invoke } from '@tauri-apps/api/core';
import { encodeData, decodeData } from './encryptedStorage';

const SERVICE = 'tech-demand-matching';
const FALLBACK_PREFIX = 'ENC:';
export type SecretStorageStatus = 'unchecked' | 'keychain' | 'local' | 'unavailable';
const statuses = new Map<string, SecretStorageStatus>();

function isTauriEnv(): boolean {
  return typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window;
}

async function tauriSave(provider: string, apiKey: string): Promise<void> {
  if (apiKey === '') {
    await invoke('delete_secret', { provider });
    return;
  }
  await invoke('save_secret', { provider, apiKey });
}

async function tauriGet(provider: string): Promise<string | null> {
  const value = await invoke<string | null>('get_secret', { provider });
  return value ?? null;
}

function localSave(provider: string, apiKey: string): void {
  if (apiKey === '') {
    localStorage.removeItem(`${SERVICE}:${provider}`);
    return;
  }
  localStorage.setItem(`${SERVICE}:${provider}`, encodeData(apiKey));
}

function localGet(provider: string): string | null {
  const raw = localStorage.getItem(`${SERVICE}:${provider}`);
  if (!raw) return null;
  if (raw.startsWith(FALLBACK_PREFIX)) return decodeData(raw) || null;
  // 兼容明文旧数据
  return raw;
}

export const secretStore = {
  async save(provider: string, apiKey: string): Promise<void> {
    if (isTauriEnv()) {
      try {
        await tauriSave(provider, apiKey);
        localStorage.removeItem(`${SERVICE}:${provider}`);
        statuses.set(provider, 'keychain');
        return;
      } catch {
        statuses.set(provider, 'unavailable');
        throw new Error('系统密钥库保存失败，密钥未写入本地兼容存储。请检查系统凭据服务后重试。');
      }
    }
    localSave(provider, apiKey);
    statuses.set(provider, 'local');
  },

  async get(provider: string): Promise<string | null> {
    if (isTauriEnv()) {
      try {
        const v = await tauriGet(provider);
        if (v !== null) {
          statuses.set(provider, 'keychain');
          return v;
        }
        const legacy = localGet(provider);
        statuses.set(provider, legacy ? 'local' : 'keychain');
        return legacy;
      } catch {
        const legacy = localGet(provider);
        statuses.set(provider, legacy ? 'local' : 'unavailable');
        if (legacy) return legacy;
        throw new Error('系统密钥库读取失败，请检查系统凭据服务后重试。');
      }
    }
    statuses.set(provider, 'local');
    return localGet(provider);
  },

  async delete(provider: string): Promise<void> {
    if (isTauriEnv()) {
      try {
        await invoke('delete_secret', { provider });
      } catch {
        statuses.set(provider, 'unavailable');
        throw new Error('系统密钥库删除失败，原密钥仍保留，请重试。');
      }
    }
    localStorage.removeItem(`${SERVICE}:${provider}`);
    statuses.set(provider, isTauriEnv() ? 'keychain' : 'local');
  },

  getStatus(provider: string): SecretStorageStatus {
    return statuses.get(provider) || (isTauriEnv() ? 'unchecked' : 'local');
  },
};

File: src-tauri/src/secret_store.rs
// OS Keychain 桥接 - 通过 keyring crate 写入系统密钥库
//
// Windows: DPAPI（Credential Manager）
// macOS:   Keychain
// Linux:   Secret Service (libsecret)
//
// key 格式：tech-demand-matching:<provider>
// value:  provider 对应的 API Key（明文，由 OS 加密落盘）

use keyring::Entry;
use serde::Serialize;

const SERVICE: &str = "tech-demand-matching";

fn make_entry(provider: &str) -> Result<Entry, String> {
    Entry::new(SERVICE, provider).map_err(|e| format!("Keychain 不可用: {e}"))
}

#[tauri::command]
pub fn save_secret(provider: String, api_key: String) -> Result<(), String> {
    if api_key.is_empty() {
        return delete_secret(provider);
    }
    let entry = make_entry(&provider)?;
    entry
        .set_password(&api_key)
        .map_err(|e| format!("保存密钥失败: {e}"))
}

#[tauri::command]
pub fn get_secret(provider: String) -> Result<Option<String>, String> {
    let entry = make_entry(&provider)?;
    match entry.get_password() {
        Ok(value) => Ok(Some(value)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(format!("读取密钥失败: {e}")),
    }
}

#[tauri::command]
pub fn delete_secret(provider: String) -> Result<(), String> {
    let entry = make_entry(&provider)?;
    match entry.delete_password() {
        Ok(()) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(format!("删除密钥失败: {e}")),
    }
}

#[derive(Serialize)]
pub struct HasSecretResult {
    pub has_secret: bool,
    pub backend: &'static str,
}

#[tauri::command]
pub fn has_secret(provider: String) -> Result<HasSecretResult, String> {
    let entry = make_entry(&provider)?;
    let has_secret = match entry.get_password() {
        Ok(_) => true,
        Err(keyring::Error::NoEntry) => false,
        Err(e) => return Err(format!("探测密钥失败: {e}")),
    };
    Ok(HasSecretResult {
        has_secret,
        backend: std::env::consts::OS,
    })
}
